### PR TITLE
Fixed model saving and loading of OneVersusAllTrainer to include SoftMax

### LINKED ISSUE #3647: LightGBM Save/Load round trip loses Softmax OneVersusAllModelParameters
version: 0.11

Related: #1424

When training a softmax Multi classifier with LightGBM a save/load will lose the softmax it seems:

Also inspecting the model we can see it doesn't use `ImplSoftmax` but `ImplRaw`.

Reproduce:

```C#
    public class GenericSample
    {
      public string A { get; set; }
      public string Label { get; set; }
    }
    public static void ReproduceLightGbmPersistanceBug()
    {
      var data = Enumerable.Range(1, 100).Select(x => new GenericSample { A = $"{x % 20}", Label = $"{x % 10}" });
      var ctx = new MLContext();

      var options = new Options {
        UseSoftmax = true,
      };
      var pipe = ctx.Transforms.Categorical.OneHotEncoding("A")
        .Append(ctx.Transforms.Concatenate("Features", "A"))
        .Append(ctx.Transforms.Conversion.MapValueToKey("Label"))
        .Append(ctx.MulticlassClassification.Trainers.LightGbm(options));
      var dataView = ctx.Data.LoadFromEnumerable(data);
      ITransformer model = pipe.Fit(dataView);
      var scores = model.Transform(dataView).GetColumn<float[]>(ctx,"Score");

      Console.WriteLine($"Min: {scores.Select(x => x.Min()).Min()}");
      Console.WriteLine($"Max: {scores.Select(x => x.Max()).Max()}");

      var memoryStream = new MemoryStream();
      ctx.Model.Save(model, memoryStream);
      model = ctx.Model.Load(memoryStream);

      scores = model.Transform(dataView).GetColumn<float[]>(ctx,"Score");
      Console.WriteLine($"Min: {scores.Select(x => x.Min()).Min()}");
      Console.WriteLine($"Max: {scores.Select(x => x.Max()).Max()}");
    }
```

Output:

```
Min: 0.001027671
Max: 0.9907509
Min: -4.843706
Max: 2.027462
```

### LINKED ISSUE #4051: LightGBM is producing different multiclass scores after loading saved model
### System information

Windows 10
Microsoft.ML (1.2.0)
Microsoft.ML.LightGbm (1.2.0)
.NET Core 2.2

### Issue

After training a lightgbm model, the model is producing multiclass scores between [0, 1] which totals 1, as expected. 

However, after saving the model, then loading it into a new trainedModel object - the scores are now not probabilities, but decimal values. 

I have tested the saving and loading with other model types and I cannot replicate the results. It is only the case with the lightgbm model.

Please advise. I am now attempting to rollback library versions to see if it's still an issue

### Source code / logs

**Before saving model...**
0.003305528
0.01293249
0.01907223
0.9646355
5.421485E-05
3.556848E-08

**After saving model...**
-3.623514
-2.259367
-1.870877
2.05264
-7.733911
-15.06316

image

Source code

mlContext.Model.Save(trainedModel, dataView.Schema, _modelPath);

            // Save Data Prep transformer
            //mlContext.Model.Save(pipeline, dataView.Schema, "data_preparation_pipeline.zip");

            schema = dataView.Schema;

            Console.WriteLine("Before saving model...");
            TestModelOutput(mlContext, trainedModel);

            // Load trained model
            trainedModel = mlContext.Model.Load(_modelPath, out schema);
            //trainedModel = mlContext.Model.LoadWithDataLoader()

            Console.WriteLine("After saving model...");
            TestModelOutput(mlContext, trainedModel);

 private static void TestModelOutput(MLContext mlContext, ITransformer model)
        {
IDataView batchData = mlContext.Data.LoadFromEnumerable(testActions);

            IDataView predictions = model.Transform(batchData);

            IEnumerable<PredictionData> predictedResults = mlContext.Data
                .CreateEnumerable<PredictionData>(predictions, reuseRowObject: false);

            foreach (var item in predictedResults)
            {
                foreach (var score in item.Score)
                {
                    Console.WriteLine(score);
                }
                
            }
}

### LINKED ISSUE #4450: LightGBMMulticlass doesn't normalize its Score column
when I try AutoML v0.16.0 multi-classification, sometimes I get prediction score <0, and the summation of all scores doesn't equal to 1. which is unexpected. Maybe some trainer's output doesn't normalize.

image

I publish a mini-reproducible project in this [repo](https://github.com/LittleLittleCloud/AutoMLMultiClassificationBug) for facilitating debug.

### PR DESCRIPTION
**Pull request by harishsk**
When the model parameters from OnVersusAllTrainer are persisted, it currently writes a single bool byte to indicate what type of output formula to use during prediction. If that byte is false, it outputs raw values else it outputs probability values normalized to one. 
The sample code in this bug uses AutoML to request a softmax output. The code currently only checks whether the output type is ImplDist which excludes softmax output.

This PR changes the saved model format in a backward compatible way by continuing to write a single byte for output type and continues to use 0 for raw output, 1 for probability normalization but also adds 2 to indicate SoftMax. These values are already in alignment with the OutputFormula enum.

Fixes #4450 
Fixes #3647
Fixes #4051

### CODE REVIEW

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs [hunk 1]
```diff
@@ -364,23 +364,29 @@ private OneVersusAllModelParameters(IHostEnvironment env, ModelLoadContext ctx)
                 : base(env, RegistrationName, ctx)
         {
             // *** Binary format ***
-            // bool: useDist
+            // byte: OutputFormula as byte
             // int: predictor count
-            bool useDist = ctx.Reader.ReadBoolByte();
+            OutputFormula outputFormula = (OutputFormula)ctx.Reader.ReadByte();
```
Comment (justinormont) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs]: Very elegant reuse of the byte.

File: test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs [hunk 2]
```diff
@@ -308,7 +308,9 @@ public void LightGbmMulticlassEstimatorWithOptions()
         {
             var options = new LightGbmMulticlassTrainer.Options
             {
-                EvaluationMetric = LightGbmMulticlassTrainer.Options.EvaluateMetricType.Default
+                EvaluationMetric = LightGbmMulticlassTrainer.Options.EvaluateMetricType.Default,
+                UseSoftmax = true
```
Comment (justinormont) [inline test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs]: You may want two unit tests total; both setting it explicitly as `true` and `false`. I believe the hyerparameter defaults to `<auto>`, where LightGBM decides for you.

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs (hunk 1 above)
Comment (harishsk) [reply src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs]: Thank you !

---
In reply to: [345483370](https://github.com/dotnet/machinelearning/pull/4472#discussion_r345483370) [](ancestors = 345483370)

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs [hunk 3]
```diff
@@ -364,23 +364,29 @@ private OneVersusAllModelParameters(IHostEnvironment env, ModelLoadContext ctx)
                 : base(env, RegistrationName, ctx)
         {
             // *** Binary format ***
-            // bool: useDist
+            // byte: OutputFormula as byte
```
Comment (ganik) [inline src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs]: You are changing how model is saved / loaded from disk. For backward compatibility you will need to use versioning

File: test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs (hunk 2 above)
Comment (harishsk) [reply test/Microsoft.ML.Tests/TrainerEstimators/TreeEstimators.cs]: Fixed. Thanks!

---
In reply to: [345950381](https://github.com/dotnet/machinelearning/pull/4472#discussion_r345950381) [](ancestors = 345950381)

File: src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs (hunk 3 above)
Comment (harishsk) [reply src/Microsoft.ML.StandardTrainers/Standard/MulticlassClassification/OneVersusAllTrainer.cs]: Backward compatibility isnt affected. In the old models, it stored a bool value as a byte. (0 = false = RawScores and 1 = true = Probability Normalization)
All I have done here is to add 2 = SoftMax. 

The old models will continue to be read and function as before. 


---
In reply to: [346003757](https://github.com/dotnet/machinelearning/pull/4472#discussion_r346003757) [](ancestors = 346003757)
